### PR TITLE
Modifying assembly scanning to only accept exclusions.

### PR DESCRIPTION
**Pull request by johnsimons**
In this new scanning we are doing the following:

1. We scan all assemblies by default (only the ones that can be loaded);
1. We provide an API to add exclusions by assembly name or `Type`:

    ```c#
    config.ExcludeAssemblies(assemblyToExclude);
    config.ExcludeTypes(typeToExclude);
    ```
1. We only include assemblies that reference the core (**see explanation below**);
1. Exclusions on assembly names are now **not using wildcard**;

### Include assemblies that reference the core
The way this works is quite simple, we check what references an assembly has, and if it includes a reference to the core, we include that assembly types.
Once that is done we walk all the types looking for `IHandleMessages<T>` and ensure that whatever the T type is, that T is also included in the types scanned, this is required for unobtrusive assemblies.

### CODE REVIEW

**Comment by johnsimons** [comment]
@SimonCropp this is starting to take shape, do u want to start looking at this and suggest ways to testing it ?

**Comment by johnsimons** [comment]
@Particular/nservicebus please review ?

**Comment by andreasohlund** [comment]
>Once that is done we walk all the types looking for IHandleMessages<T> and ensure that whatever the T type is, that T is also included in the types scanned, this is required for unobtrusive assemblies.

Does this means that we won't include message assemblies using unobtrusive? (since they won't reference the core)

**Comment by SimonCropp** [comment]
@andreasohlund yes

**Comment by andreasohlund** [comment]
So a user would have to include them explicitly?

**Comment by SimonCropp** [comment]
i believe the intent is to include the message on first use

**Comment by johnsimons** [comment]
No that is not it, the message unobtrusive message types are also included.
Obviously my explanation was not clear enough.

**Comment by danielmarbach** [comment]
@johnsimons and @SimonCropp I'm doing heavy changes in the MetaDataRegistry as well. How can we bring those two together efficiently?

File: src/NServiceBus.AcceptanceTesting/Customization/BusConfigExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+namespace NServiceBus.AcceptanceTesting.Customization
+{
+    using System;
+    using System.Collections.Generic;
+
+    public static class BusConfigExtensions
+    {
+        /// <summary>
+        /// Backdoor into the core types to scan. This allows people to test a subset of functionality when running Acceptance tests
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTesting/Customization/BusConfigExtensions.cs]: How does the statement in the comment hold true when we do not release the ATT library? what can people do who are not using it?

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs [hunk 2]
```diff
@@ -17,12 +18,6 @@ public enum AddressMode
         Local = 0,
         Remote = 1,
     }
-    public class AllAssemblies : NServiceBus.IExcludesBuilder, NServiceBus.IIncludesBuilder, System.Collections.Generic.IEnumerable<System.Reflection.Assembly>, System.Collections.IEnumerable
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Shouldn't that be properly obsoleted as well?

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs [hunk 3]
```diff
@@ -365,19 +366,10 @@ public interface IDataBusProperty
         void SetValue(object value);
     }
     public interface IEvent : NServiceBus.IMessage { }
-    public interface IExcludesBuilder : System.Collections.Generic.IEnumerable<System.Reflection.Assembly>, System.Collections.IEnumerable
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Obsoletion?

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs [hunk 4]
```diff
@@ -365,19 +366,10 @@ public interface IDataBusProperty
...
-    {
-        NServiceBus.IExcludesBuilder And(string assemblyExpression);
-    }
     public interface IHandleMessages<T>
     {
         void Handle(T message);
     }
-    public interface IIncludesBuilder : System.Collections.Generic.IEnumerable<System.Reflection.Assembly>, System.Collections.IEnumerable
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Obsoletion?

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs [hunk 5]
```diff
@@ -2244,7 +2239,6 @@ namespace NServiceBus.Unicast.Routing
     public class StaticMessageRouter
     {
         public StaticMessageRouter(System.Collections.Generic.IEnumerable<System.Type> knownMessages) { }
-        public bool SubscribeToPlainMessages { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Obsoletion?

File: src/NServiceBus.Core.Tests/AssemblyScanner/When_directory__with_non_dot_net_dll_is_scanned.cs [hunk 6]
```diff
@@ -17,17 +14,10 @@ public class When_directory__with_non_dot_net_dll_is_scanned
...
         public void Context()
         {
-            var codeBase = Assembly.GetExecutingAssembly().CodeBase;
-            var uri = new UriBuilder(codeBase);
-            var path = Uri.UnescapeDataString(uri.Path);
-            var directoryName = Path.GetDirectoryName(path);
-
-            var testDllDirectory = Path.Combine(directoryName, "TestDlls");
-            var assemblyScanner = new AssemblyScanner(testDllDirectory)
+            var assemblyScanner = new AssemblyScanner(AssemblyScannerTests.GetTestAssemblyDirectory())
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/AssemblyScanner/When_directory__with_non_dot_net_dll_is_scanned.cs]: The containing class has two underscores

Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/AssemblyScanner/When_directory__with_non_dot_net_dll_is_scanned.cs]: I'm always very sceptic when I see a test which references another test. I know this is nitpicker corner but this usually fires an alarm in my reviewer brain ;) I know in that case this isn't really an issue but I want to point it out.

File: src/NServiceBus.Core.Tests/AssemblyScanner/When_directory_with_no_reference_dlls_is_scanned.cs [hunk 7]
```diff
@@ -16,17 +13,10 @@ public class When_directory_with_no_reference_dlls_is_scanned
...
         public void Context()
         {
-            var codeBase = Assembly.GetExecutingAssembly().CodeBase;
-            var uri = new UriBuilder(codeBase);
-            var path = Uri.UnescapeDataString(uri.Path);
-            var directoryName = Path.GetDirectoryName(path);
-
-            var testDllDirectory = Path.Combine(directoryName, "TestDlls");
-            var assemblyScanner = new AssemblyScanner(testDllDirectory)
+            var assemblyScanner = new AssemblyScanner(AssemblyScannerTests.GetTestAssemblyDirectory())
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/AssemblyScanner/When_directory_with_no_reference_dlls_is_scanned.cs]: See https://github.com/Particular/NServiceBus/pull/2670/files#r26283550

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs (hunk 5 above)
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Side note: I've pushed a change to apicomparer that lets you compare your "c./dev/NServiceBus/Binaries/NServiceBus.Core.dll" against the latest on the build server.

That call creates a nice report on whats been changed that at least for me is invaluable to detect and put the needed obsoletes in

https://github.com/ParticularLabs/APIComparer/commit/e99bcac6da34755a181408fbdeac18bf8b84b061

File: src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj [hunk 8]
```diff
@@ -297,21 +294,27 @@
...
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </Content>
-    <Content Include="TestDlls\libzmq-v120-mt-3_2_3.dll">
+    </None>
+    <None Include="TestDlls\AssemblyWithReference.dll">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </Content>
-    <Content Include="TestDlls\some_random.dll">
+    </None>
+    <None Include="TestDlls\dotNet.dll">
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj]: Could we somehow prefix the assemblies with the test they belong? When I come back in a few months and see dotNet.dll or some_random.dll I might not remember why they are here and what they are doing

File: src/NServiceBus.Core/AllAssemblies.cs [hunk 9]
```diff
@@ -1,115 +0,0 @@
...
-    using System.Collections;
-    using System.Collections.Generic;
-    using System.Reflection;
-    using System.Web;
-    using Hosting.Helpers;
-
-    /// <summary>
-    /// Class for specifying which assemblies not to load.
-    /// </summary>
-    public class AllAssemblies : IExcludesBuilder, IIncludesBuilder
```
Comment (danielmarbach) [inline src/NServiceBus.Core/AllAssemblies.cs]: Obsoletion?

File: src/NServiceBus.Core/DataBus/DataBus.cs [hunk 10]
```diff
@@ -56,43 +51,5 @@ protected internal override void Setup(FeatureConfigurationContext context)
             context.Pipeline.Register<DataBusReceiveBehavior.Registration>();
             context.Pipeline.Register<DataBusSendBehavior.Registration>();
         }
-
```
Comment (danielmarbach) [inline src/NServiceBus.Core/DataBus/DataBus.cs]: Can you explain why this was removed in that PR?

File: src/NServiceBus.Core/DataBus/DataBusFileBased.cs [hunk 11]
```diff
@@ -15,21 +15,14 @@ public DataBusFileBased()
...
         {
-            // We still doing this check here eventhough we now have an explicit API
-            // to register custom IDataBus implementations for backwards compatibility
-            var customDataBusComponentRegistered = context.Container.HasComponent<IDataBus>();
-
-            if (!customDataBusComponentRegistered)
+            string basePath;
+            if (!context.Settings.TryGet("FileShareDataBusPath", out basePath))
             {
-                string basePath;
```
Comment (danielmarbach) [inline src/NServiceBus.Core/DataBus/DataBusFileBased.cs]: Can you explain why this was removed in that PR?

File: src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs [hunk 12]
```diff
@@ -316,32 +433,20 @@ bool IsIncluded(string assemblyNameOrFileName)
...
 
-            var noAssembliesWereExplicitlyIncluded = !AssembliesToInclude.Any();
-            var isExplicitlyIncluded = AssembliesToInclude.Any(included => IsMatch(included, assemblyNameOrFileName));
-
-            return noAssembliesWereExplicitlyIncluded || isExplicitlyIncluded;
+            return true;
         }
 
-        static bool IsMatch(string expression, string scopedNameOrFileName)
+        static bool IsMatch(string expression1, string expression2)
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs]: Very descriptive parameter name :)

File: src/NServiceBus.Core/BusConfiguration.cs [hunk 13]
```diff
@@ -61,36 +60,71 @@ public void RegisterComponents(Action<IConfigureComponents> registration)
...
+            if (assemblies == null)
+            {
+                throw new ArgumentNullException("assemblies");
+            }
+
+            if (assemblies.Any(string.IsNullOrWhiteSpace))
+            {
+                throw new ArgumentException("Passed in a null or empty assembly name.", "assemblies");
+            }
+            excludedAssemblies = excludedAssemblies.Union(assemblies, StringComparer.OrdinalIgnoreCase).ToList();
```
Comment (danielmarbach) [inline src/NServiceBus.Core/BusConfiguration.cs]: Cool so this makes it additive. Nice. Therefore we can now tackle the issues with the testing framework and xunit exclusions, right?

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs (hunk 5 above)
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: @andreasohlund You ApiComparer Marketing minister you! ;)

File: src/NServiceBus.Core/DataBus/DataBusFileBased.cs (hunk 11 above)
Comment (johnsimons) [inline src/NServiceBus.Core/DataBus/DataBusFileBased.cs]: Can;t really remember

File: src/NServiceBus.Core/DataBus/DataBus.cs (hunk 10 above)
Comment (johnsimons) [inline src/NServiceBus.Core/DataBus/DataBus.cs]: Not needed anymore, databus is now explicitly activated

File: src/NServiceBus.AcceptanceTesting/Customization/BusConfigExtensions.cs (hunk 1 above)
Comment (johnsimons) [inline src/NServiceBus.AcceptanceTesting/Customization/BusConfigExtensions.cs]: Not sure I totally understand the question, anyway I'll try to answer it.
This is a hack to support our AT framework.
Users should not require this hack.

**Comment by johnsimons** [comment]
I have added the missing obsoletes.
Any other issues or is ok to merge ?

**Comment by SeanFeldman** [comment]
@johnsimons, would it also benefit to add an ability to exclude folders? A few times was asked in the [user group](https://groups.google.com/forum/#!topic/particularsoftware/7Me1KKAxmFU)
```c#
config.ExcludeFolders(foldersToExclude[]);
```

File: src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: @danielmarbach isnt it obsoleted on the line below?

Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/approvals/NServiceBus.Core.approved.cs]: Hmmm... Must have missed that

File: src/NServiceBus.Core/BusConfiguration.cs [hunk 14]
```diff
@@ -61,36 +60,71 @@ public void RegisterComponents(Action<IConfigureComponents> registration)
         }
 
         /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="assemblies"></param>
+        public void ExcludeAssemblies(params string[] assemblies)
```
Comment (SimonCropp) [inline src/NServiceBus.Core/BusConfiguration.cs]: is an empty list valid here?

Comment (johnsimons) [inline src/NServiceBus.Core/BusConfiguration.cs]: Yes sure, but not sure what that achieves?

**Comment by johnsimons** [comment]
@SeanFeldman it may be beneficial to add an exclude folder but that is already possible to accompish with the api we are adding.

**Pull request merged**